### PR TITLE
improvement(core): make module resolution faster to reduce startup time

### PR DESCRIPTION
**Pull request by edvald**
**What this PR does / why we need it**:

This makes module template string resolution happen more explicitly
in dependency order, and avoids repeated resolution of the same module
while resolving template strings. Avoids issues with large/complex
projects, such as slow startups.

**Special notes for your reviewer**:

This is a fairly substantial refactor. Please look for any potentially missing test cases, and test manually if you can think of potential issues.

### CODE REVIEW

File: garden-service/test/unit/src/task-graph.ts [hunk 1]
```diff
@@ -542,13 +542,13 @@ describe("task-graph", () => {
       const t2 = new TestTask(garden, "a", false, { uid: "2", versionString: "2", callback: repeatedCallback("2") })
       const t3 = new TestTask(garden, "a", false, { uid: "3", versionString: "3", callback: repeatedCallback("3") })
 
-      const firstProcess = graph.process([t1])
+      const firstProcess = await graph.process([t1])
```
Comment (thsig) [inline garden-service/test/unit/src/task-graph.ts]: Is there a reason we're adding `await`s in these test cases? They were meant to test concurrent calls to `graph.process`.

File: garden-service/src/tasks/resolve-module.ts [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+          chalk.red(
+            `Could not find build dependency ${chalk.white(name)}, configured in module ${chalk.white(
+              this.moduleConfig.name
+            )}`
+          ),
+          { moduleConfig }
+        )
+      }
+
+      if (d[2] === "version") {
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/tasks/resolve-module.ts]: I agree, I'll add a comment

File: garden-service/test/unit/src/task-graph.ts (hunk 1 above)
Comment (edvald) [reply garden-service/test/unit/src/task-graph.ts]: Oops, that's a refactoring mistake, I'll revert that.

**Comment by edvald** [comment]
@thsig comments addressed, should be good to go

**Comment by eysi09** [comment]
@edvald, we need to resolve conflicts before merging.

**Comment by edvald** [comment]
I also got some feedback on this on the community Slack. I'm gonna test a user project with this branch, it seems like there is still something that needs to be fixed.

**Comment by edvald** [comment]
@thsig this needs additional review, but should be good to go (after rebase + squash). It's been verified by @swist to solve their problem.

**Pull request merged**